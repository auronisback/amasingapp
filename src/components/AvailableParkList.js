import React, {Component} from 'react';
import ParkStroke from "./ParkStroke";

/**
 * Component managing the set of available parking lots.
 * */
export default class AvailableParkList extends Component {

  /**
   * Strokes appearance.
   * */
  static STROKE_COLOR = '#777';
  static STROKE_OPACITY = 0.5;

  /**
   * Renders the component.
   * @return The react component
   * */
  render() {
    return (
        <div>
          {this._buildAvailableParks()}
        </div>
    );
  }

  /**
   * Creates elements for the available parking lots.
   * */
  _buildAvailableParks = () => {
    let availableParks = [];
    if(this.props.parks)
      this.props.parks.forEach((elem) => {
        availableParks.push(
            <ParkStroke
                key={elem.id}
                start={elem.start}
                end={elem.end}
                color={AvailableParkList.STROKE_COLOR}
                opacity={AvailableParkList.STROKE_OPACITY}
            />);
      });
    return availableParks;
  }
}