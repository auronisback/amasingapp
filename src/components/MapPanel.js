import React, {Component} from 'react';
import {withScriptjs, withGoogleMap, GoogleMap} from 'react-google-maps';

/**
 * Wrapper for the Google Map.
 * */
const MapWrapper = withScriptjs(withGoogleMap((props) => (
    <GoogleMap
        center={props.center}
        defaultZoom={props.zoom}
        onClick={props.onMapClick}
    >
      {props.children}
    </GoogleMap>
)));

/**
 * MapPanel Component
 * Manages the rendering of the panel which shows the
 * Google Maps map.
 * */
export class MapPanel extends Component {

  /**
   * Rendering method, displays the component.
   *
   * @returns The rendered component
   * */
  render() {
    const {center, zoom, gMapsApiKey} = this.props;
    return (<div>
      <MapWrapper
          googleMapURL={'https://maps.googleapis.com/maps/api/js?key=' + gMapsApiKey + '&v=3.exp&libraries=geometry,drawing,places'}
          loadingElement={<div style={{height: '100px'}}/>}
          containerElement={<div className={'MapPanel'}/>}
          mapElement={<div style={{height: '100%'}}/>}
          center={center}
          zoom={zoom}
          onMapClick={this.onMapClick}
      >
        {this.props.children}
      </MapWrapper>
    </div>);
  }

  /**
   * Handler for click on the map. Calls the destination setter of
   * with the handler specified in props.
   * @param {Event} evt The generated map event
   * */
  onMapClick = (evt) => {

    if(this.props.onDestinationSelection)
      this.props.onDestinationSelection(evt.latLng.lat(), evt.latLng.lng());
  };

}