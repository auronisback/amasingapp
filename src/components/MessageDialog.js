import React, {Component} from 'react';

const NO_RESULTS = 'Server returned status code ZERO_RESULTS';

export class MessageDialog extends Component {

  /**
   * Constants for error types.
   * */
  static TYPE_MESSAGE = 'info';
  static TYPE_WARNING = 'warning';
  static TYPE_ERROR = 'danger';

  /**
   * Renderizes the component.
   * @return The React Component
   * */
  render() {
    let message = this.getMessage();
    const typeClass = 'alert-' + this.props.type;
    return (
        <div className={'alert ' + typeClass + ' ErrorDialog'}
             onClick={this.props.onClose}
        >
          <h4 className={'alert-heading'}>Warning!</h4>
          <p>{message}</p>
          <span className={'CloseError'}>Click to close</span>
        </div>
    );
  }

  /**
   * Gets the error message from component's properties.
   * @return string The error message in human readable form
   * */
  getMessage = () => {
    switch(this.props.message) {
      case NO_RESULTS:
        return 'The search produced no results.';
      default:
        return this.props.message;
    }
  };

}