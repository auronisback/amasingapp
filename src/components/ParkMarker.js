import React, {Component} from 'react';
import {InfoWindow, Marker} from "react-google-maps";

import ParkInfo from './ParkInfo';
import ParkStroke from "./ParkStroke";


/**
 * Component used to manage parking lots markers.
 * */
export class ParkMarker extends Component {

  /**
   * Component's state.
   * */
  state = {
    infoWindowShown: false
  };

  /**
   * Strokes appearance.
   * */
  static STROKE_COLOR = '#F00';
  static STROKE_OPACITY = 1;

  /**
   * Renders the component.
   * */
  render() {
    return (
        <Marker
            position={this.getPosition()}
            onClick={this.toggleOpen}
            label={'' + this.props.label}
        >
          {
            this.state.infoWindowShown && (<InfoWindow
                onCloseClick={this.handleClose}
            >
              <ParkInfo
                  occRate={this.props.parking.occupancy}
                  distance={this.props.parking.distance}
                  parkNumber={this.props.parking.number}
                  streetType={this.props.parking.streetType}
                  haveMes={this.props.parking.haveMes}
              />
            </InfoWindow>)
          }
          <ParkStroke
              start={this.props.parking.start}
              end={this.props.parking.end}
              color={ParkMarker.STROKE_COLOR}
              opacity={ParkMarker.STROKE_OPACITY}
          />
        </Marker>
    );
  };

  /**
   * Handler for the click on the marker, used to show
   * the info window.
   * */
  toggleOpen = () => {
    this.setState({infoWindowShown: !this.state.infoWindowShown})
  };

  /**
   * Handler for closing the info window.
   * */
  handleClose = () => {
    this.setState({infoWindowShown: false})
  };

  /**
   * Gets the position of the marker.
   * @return {Object} An object with coordinates for the marker
   * */
  getPosition = () => {
    return {
      lat: (this.props.parking.start.lat + this.props.parking.end.lat) / 2,
      lng: (this.props.parking.start.lng + this.props.parking.end.lng) / 2
    };
  };
}