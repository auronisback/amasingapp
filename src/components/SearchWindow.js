import React, {Component} from 'react';
import Slider from 'rc-slider';
import {library} from '@fortawesome/fontawesome-svg-core';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faWalking, faBlind, faSpinner} from "@fortawesome/free-solid-svg-icons";
import 'rc-slider/assets/index.css';

export class SearchWindow extends Component {

  /**
   * Creates the object, initializing icons.
   * @param {Object} props React component's properties
   * */
  constructor(props) {
    super(props);
    library.add(faWalking, faBlind, faSpinner);
  }

  /**
   * Renders the object.
   * @return Component The React component
   * */
  render() {
    // Rendering component
    return (
        <form onSubmit={this.onFormSubmitted}>
          <div className={'SearchPanel form-group'}>
            <select className={'form-control'}
                    value={this.props.dow}
                    onChange={this.onDowChanged}
            >
              <option value={0}>Sunday</option>
              <option value={1}>Monday</option>
              <option value={2}>Tuesday</option>
              <option value={3}>Wednesday</option>
              <option value={4}>Thursday</option>
              <option value={5}>Friday</option>
              <option value={6}>Saturday</option>
            </select>
            <input type={'time'} className={'form-control'}
                   value={this.props.time.h + ':' +
                   ('' + this.props.time.m).padStart(2, '0')}
                   onChange={this.onTimeChanged}
            />
            <input type={'number'} placeholder={'distance (m)'} className={'form-control'}
                   value={this.props.distance === 0 ? '' : this.props.distance}
                   onChange={this.onDistanceChanged}
                   min={0}
                   step={50}
            />
            <div className={'WalkingSlider'}>
              <div className={'IconWrapper'}>
                <FontAwesomeIcon icon={"blind"}/>
              </div>
              <Slider className={'Slider'}
                      onChange={this.onSliderChanged}
                      value={this.props.walkWeight}
              />
              <div className={'IconWrapper'}>
                <FontAwesomeIcon icon={"walking"}/>
              </div>
            </div>
            <button className={'btn btn-primary btn-sm'}
                    disabled={this.props.isLoading}
            >
              {this.props.isLoading ? (<FontAwesomeIcon icon={'spinner'} pulse/>) : 'Search'}
            </button>
          </div>
        </form>
    );
  }

  /**
   * Handler method for the change on dow.
   *
   * @param {Event} evt The generated event
   * */
  onDowChanged = (evt) => {
    this.setState({dow: evt.target.value});
  };

  /**
   * Handler method for the change on time.
   *
   * @param {Event} evt The generated event
   * */
  onTimeChanged = (evt) => {
    const time = SearchWindow.getTimeFromString(evt.target.value);
    // Calling the handler function
    this.props.onTimeChanged({
        h: time.getHours(),
        m: time.getMinutes()
    });
  };

  /**
   * Handler function for changing in the distance input.
   *
   * @param {Event} evt The generated event
   * */
  onDistanceChanged = (evt) => {
    this.props.onDistanceChanged(evt.target.value);
  };

  /**
   * Handler for slider changing.
   *
   * @param {number} value Slider value
   * */
  onSliderChanged = (value) => {
    this.props.onWalkWeightChanged(value);
  };

  /**
   * Handler function for submitting the form.
   *
   * @param {Event} evt The generated event
   * */
  onFormSubmitted = (evt) => {
    evt.nativeEvent.preventDefault(); // Do not actually submit
    if(this.props.onSubmit)
      this.props.onSubmit()
  };

  /**
   * Static utility function to retrieve Date object from
   * time inputs.
   *
   * @param {Object} time The time input as a string with format 'HH:mm'
   * */
  static getTimeFromString(time) {
    let pieces = time.split(':');
    time = new Date();
    time.setHours(pieces[0]);
    time.setMinutes(pieces[1]);
    return time;
  };

}