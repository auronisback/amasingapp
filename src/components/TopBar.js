import React, {Component} from 'react';
import {library} from '@fortawesome/fontawesome-svg-core';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faSearch, faUndo} from "@fortawesome/free-solid-svg-icons";


export class TopBar extends Component {

  /**
   * Component's state.
   * */
  state = {
    streetName: ''
  };

  constructor(props) {
    super(props);
    // Initializing font-awesome icons
    library.add(faSearch, faUndo);
  }

  render() {
    return (
        <header className={'container fluid TopBar'}>
          <div className={'row'}>
            <div className={'col-6'}>
              <h1>aMasIngApp</h1>
            </div>
            <div className={'col-6'}>
              <ul>
                <li>
                  <button className={'btn btn-primary ResetButton'}
                          onClick={this.props.onReset}>
                    <FontAwesomeIcon icon={'undo'} />
                  </button>
                </li>
                <li>
                  <div className={'input-group'}>
                    <input type={'text'} className={'form-control'} placeholder={'Search by street'}
                           onChange={this.onSearchInputChanged}
                           value={this.state.streetName}
                           onKeyPress={this.onInputKeyPressed}
                    />
                    <button className={'input-group-append btn btn-primary SearchButton'}
                            onClick={this.onSearch}
                    >
                      <FontAwesomeIcon icon={"search"}/>
                    </button>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </header>
    );
  }

  /**
   * Action to be performed when the search button is pressed. The action
   * will be performed if the input has at least two characters.
   * */
  onSearch = () => {
    if (this.props.onSearch && this.state.streetName.length > 2)
      this.props.onSearch(this.state.streetName); // Taken from component's state
  };

  /**
   * Handler for changing the street name input.
   * @param {Event} evt The generated event
   * */
  onSearchInputChanged = (evt) => {
    this.setState({streetName: evt.target.value});
  };

  /**
   * Handler for the pressure of a key in the input. If the pressed key
   * is ENTER, then it will fire the search.
   * @param {Event} evt The generated event
   * */
  onInputKeyPressed = (evt) => {
    if(evt.key === 'Enter')
      this.onSearch();
  };
}