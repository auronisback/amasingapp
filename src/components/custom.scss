// Override default variables before the import
$body-bg: #DDD;

// Color palette
$backMain: #4484CE;
$backSecond: #F9CF00;
$foreMain: #F19F4d;
$foreSecond: #D9D9D9;

// Import Bootstrap and its default variables
@import '~bootstrap/scss/bootstrap.scss';

.App {
  height: 100vh;
  width: 100%;
  overflow: hidden;
}

.TopBar {
  background: $backSecond;
  max-width: 100%;
}

.TopBar h1 {
  font-size: 200%;
  margin: 0.6rem 0;
  font-weight: bold;
  color: $backMain;
}

.TopBar ul {
  margin-bottom: 0;
  text-align: right;
}

.TopBar li {
  display: inline-block;
  padding: 2%;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.TopBar .SearchButton {
  -webkit-border-radius: 0 0.25rem 0.25rem 0;
  -moz-border-radius: 0 0.25rem 0.25rem 0;
  border-radius: 0 0.25rem 0.25rem 0;
}

.MapPanel {
  height: 100vh;
  width: 100%;
}

.DestinationMarker {
  position: relative;
  display: inline-block;
  text-align: center;
  padding: 0.4em;
  width: 200px;
}

.DestinationMarker .Body {
  position: absolute;
}

.DestinationMarker .Body.SearchOpened {
  left: -125px;
  top: -255px;
}

.DestinationMarker .Body.SearchClosed {
  left: -11px;
  top: -30px;
}

.MarkerIcon {
  font-size: 30px;
  display: inline-block;
  text-align: center;
}

.Marker {
  color: $backMain;
}

.ParkMarker {
  margin-top: -35px;
  margin-left: -8px;
}

.SearchPanel {
  display: inline-block;
  padding: 1em;
  margin-bottom: 0.2em;
  background: $backSecond;
  -webkit-border-radius: 7px;
  -moz-border-radius: 7px;
  border-radius: 7px;
}

.SearchPanel input, .SearchPanel select {
  margin-bottom: 0.4em;
}

.SearchPanel .WalkingSlider {
  color: $backMain;
  font-size: 12pt;
  margin: 0.6em 0;
}

.SearchPanel .Slider {
  display: inline-block;
  width: 80%;
}

.SearchPanel .Slider .rc-slider-track {
  background: $backMain;
}

.SearchPanel .Slider .rc-slider-handle {
  border-color: $backMain;
}

.SearchPanel .IconWrapper {
  display: inline-block;
  width: 10%;
}

.SearchPanel .IconWrapper:first-child {
  text-align: left;
}

.SearchPanel .IconWrapper:last-child {
  text-align: right;
}

.ErrorDialog {
  cursor: pointer;
  position: fixed;
  bottom: 1em;
  right: 1em;
}

.CloseError {
  font-style: italic;
  font-size: 80%;
}