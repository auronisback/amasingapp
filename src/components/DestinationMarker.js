import React, {Component} from 'react';
import {SearchWindow} from './SearchWindow';
import {Marker, OverlayView, Circle} from "react-google-maps";

/**
 * Gets the marker's offset for the search window.
 * */
const getPixelPositionOffset = pixelOffset => (width, height) => ({
  x: -(width / 2) + pixelOffset.x,
  y: -(height / 2) + pixelOffset.y
});

/**
 * Default max distance for searches.
 * */
const DEFAULT_MAX_DISTANCE = 100;

/**
 * Component defining the marker for the destination.
 * */
export class DestinationMarker extends Component {

  /**
   * Distance circle appearance.
   * */
  static CIRCLE_FILL_COLOR = '#F00';
  static CIRCLE_STROKE_WEIGHT = 0;

  /**
   * Offset in pixel for the marker.
   * */
  static markerPixelOffset = {
    x: 0,
    y: -32
  };

  /**
   * React Component's state.
   * */
  state = {
    distance: DEFAULT_MAX_DISTANCE,
    dow: new Date().getDay(),
    time: {
      h: new Date().getHours(),
      m: new Date().getMinutes()
    },
    walkWeight: 50
  };

  /**
   *  Rendering method for React Components
   *  @return Component The react component
   * */
  render() {
    // If no destination has been set, nothing will be rendered
    if(!this.props.destination)
      return null;
    // Conditionally rendering the search window
    const overlay = (this.props.searchVisible ?
            <OverlayView
                position={this.props.destination}
                mapPaneName={OverlayView.OVERLAY_MOUSE_TARGET}
                getPixelPositionOffset={getPixelPositionOffset(DestinationMarker.markerPixelOffset)}
                options={{enableEventPropagation: false}}
            >
              <SearchWindow
                  onSubmit={this.onSubmit}
                  className={'SearchOpened'}
                  isLoading={this.props.isLoading}
                  dow={this.state.dow}
                  distance={this.state.distance}
                  time={this.state.time}
                  walkWeight={this.state.walkWeight}
                  onDowChanged={this.onDowUpdate}
                  onTimeChanged={this.onTimeUpdate}
                  onDistanceChanged={this.onDistanceUpdate}
                  onWalkWeightChanged={this.onWalkWeightUpdate}
              />
            </OverlayView> : null
    );
    return (
        <Marker position={this.props.destination}>
          {overlay}
          <Circle
              center={this.props.destination}
              radius={this.state.distance}
              options={{
                fillColor: DestinationMarker.CIRCLE_FILL_COLOR,
                strokeWeight: DestinationMarker.CIRCLE_STROKE_WEIGHT
              }}
          />
        </Marker>
    );
  }

  /**
   * Handler method when dow is changed into search window.
   * @param {int} dow Selected day of the week
   * */
  onDowUpdate = (dow) => {
    this.setState({dow: dow});
  };

  /**
   * Handler method when time is changed.
   * @param {Object} time Time in the search window
   * */
  onTimeUpdate = (time) => {
    this.setState({time: time});
  };

  /**
   * Handler method for changing in the max distance.
   * @param {string} distance Max distance for parks
   * */
  onDistanceUpdate = (distance) => {
    this.setState({distance: +distance > 0 ? +distance : 0});
  };

  /**
   * Handler method for updating the walking weight.
   * @param {number} walkWeight Weight given to walk
   * */
  onWalkWeightUpdate = (walkWeight) => {
    this.setState({walkWeight: walkWeight});
  };

  /**
   * Handles the submitting of search parameters.
   * */
  onSubmit = () => {
    if(this.props.onSearch)
      this.props.onSearch(this.state);
  };


}