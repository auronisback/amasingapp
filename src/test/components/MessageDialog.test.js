import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MessageDialog } from "../../components/MessageDialog";

Enzyme.configure({adapter: new Adapter()});

describe('<MessageDialog>', () => {
  it('Should render a message', () => {
    const message = 'Test message';
    const dialog = shallow(<MessageDialog
        message={message}
        type={MessageDialog.TYPE_MESSAGE}
    />);
    expect(dialog).toMatchSnapshot();
  });

  it('Should render a warning', () => {
    const message = 'Test warning';
    const dialog = shallow(<MessageDialog
        message={message}
        type={MessageDialog.TYPE_WARNING}
    />);
    expect(dialog).toMatchSnapshot();
  });

  it('Should render an error', () => {
    const message = 'Test error';
    const dialog = shallow(<MessageDialog
        message={message}
        type={MessageDialog.TYPE_ERROR}
    />);
    expect(dialog).toMatchSnapshot();
  });

  it('Should call the close handler function when clicked', () => {
    const message = 'Test closing';
    const closeFn = jest.fn(() => {});
    const dialog = shallow(<MessageDialog
        message={message}
        type={MessageDialog.TYPE_ERROR}
        onClose={closeFn}
    />);
    dialog.simulate('click');
    expect(closeFn).toBeCalled();
  });
});