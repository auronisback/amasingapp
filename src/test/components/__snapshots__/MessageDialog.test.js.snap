// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageDialog> Should render a message 1`] = `
<div
  className="alert alert-info ErrorDialog"
>
  <h4
    className="alert-heading"
  >
    Warning!
  </h4>
  <p>
    Test message
  </p>
  <span
    className="CloseError"
  >
    Click to close
  </span>
</div>
`;

exports[`<MessageDialog> Should render a warning 1`] = `
<div
  className="alert alert-warning ErrorDialog"
>
  <h4
    className="alert-heading"
  >
    Warning!
  </h4>
  <p>
    Test warning
  </p>
  <span
    className="CloseError"
  >
    Click to close
  </span>
</div>
`;

exports[`<MessageDialog> Should render an error 1`] = `
<div
  className="alert alert-danger ErrorDialog"
>
  <h4
    className="alert-heading"
  >
    Warning!
  </h4>
  <p>
    Test error
  </p>
  <span
    className="CloseError"
  >
    Click to close
  </span>
</div>
`;
